libs/platform/ebpf_object.c

```diff
@@ -128,26 +206,75 @@ ebpf_duplicate_utf8_string(_Out_ ebpf_utf8_string_t* destination, _In_ const ebp
     }
 }
 
+_Requires_lock_held_(&_ebpf_object_tracking_list_lock) static ebpf_object_t* _get_next_object(
+    ebpf_id_t start_id, ebpf_object_type_t object_type)
+{
+    // The start_id need not exist, so we can't call _get_index_from_id().
+    uint32_t index = (start_id >> 16);
+    if (_get_id_from_index(index) == start_id) {
+        index++;
+    }
+    while (index < EBPF_COUNT_OF(_ebpf_id_table)) {
+        ebpf_object_t* object = _ebpf_id_table[index].object;
+        if ((object != NULL) && (object->type == object_type)) {
+            return object;
+        }
+        index++;
+    }
+    return NULL;
+}
+
+ebpf_result_t
+ebpf_get_next_id(ebpf_id_t start_id, ebpf_object_type_t object_type, _Out_ ebpf_id_t* next_id)
+{
+    ebpf_result_t return_value = EBPF_NO_MORE_KEYS;
+
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
+
+    ebpf_object_t* object = _get_next_object(start_id, object_type);
+    if (object != NULL) {
+        *next_id = object->id;
+        return_value = EBPF_SUCCESS;
+    }
+
+    ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
+    return return_value;
+}
+
 void
 ebpf_object_reference_next_object(ebpf_object_t* previous_object, ebpf_object_type_t type, ebpf_object_t** next_object)
 {
     ebpf_lock_state_t state;
-    ebpf_list_entry_t* entry;
     *next_object = NULL;
 
     state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-    if (previous_object == NULL)
-        entry = _ebpf_object_tracking_list.Flink;
-    else
-        entry = previous_object->global_list_entry.Flink;
-
-    for (; entry != &_ebpf_object_tracking_list; entry = entry->Flink) {
-        ebpf_object_t* object = EBPF_FROM_FIELD(ebpf_object_t, global_list_entry, entry);
-        if (object->type == type) {
-            *next_object = object;
-            ebpf_object_acquire_reference(object);
-            break;
-        }
+
+    ebpf_id_t start_id = (previous_object) ? previous_object->id : 0;
+    ebpf_object_t* object = _get_next_object(start_id, type);
+    if (object != NULL) {
+        *next_object = object;
+        ebpf_object_acquire_reference(object);
     }
+
+    ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
+}
+
+ebpf_result_t
+ebpf_reference_object_by_id(ebpf_id_t id, ebpf_object_type_t object_type, _Outptr_ ebpf_object_t** object)
+{
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
+
+    uint32_t index;
+    ebpf_result_t return_value = _get_index_from_id(id, &index);
```

Comment:
why not have a _ebpf_get_object_by_id (since there is already _get_next_object) as well and invoke that here. Add reference at end.

Reply:
I don't follow.  That's what ebpf_reference_object_by_id() already is.

Reply:
I meant to say that just like `ebpf_object_reference_next_object` calls `_get_next_object`, maybe `ebpf_reference_object_by_id` could call a `_get_object_by_id`. The second function could then be used from other functions. Also consider renaming `_get_next_object` to `_get_next_object_by_id` and `ebpf_reference_object_by_id` to `ebpf_object_reference_by_id`

Reply:
> I meant to say that just like `ebpf_object_reference_next_object` calls `_get_next_object`, maybe `ebpf_reference_object_by_id` could call a `_get_object_by_id`. The second function could then be used from other functions. 

Why?  Right now the intent is this should be the only function to do so, so that you safely get a reference.  What other function might there be?   Is this a blocking comment?

> Also consider renaming `_get_next_object` to `_get_next_object_by_id` and `ebpf_reference_object_by_id` to `ebpf_object_reference_by_id`

Done